subnet/kube/kube.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2015 flannel authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"time"
+
+	"github.com/coreos/flannel/pkg/ip"
+	"github.com/coreos/flannel/subnet"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+var (
+	ErrUnimplemented = errors.New("unimplemented")
+	kubeSubnetCfg    *subnet.Config
+)
+
+const (
+	resyncPeriod = 5 * time.Minute
+
+	backendDataAnnotation     = "net.alpha.kubernetes.io/flannel-backend-data"
+	backendTypeAnnotation     = "net.alpha.kubernetes.io/flannel-backend-type"
+	backendPublicIPAnnotation = "net.alpha.kubernetes.io/flannel-public-ip"
+
+	cniConfPath       = "/etc/kube-flannel/cni-conf.json"
+	cniConfTargetPath = "/etc/cni/net.d/10_flannel.conf"
+	netConfPath       = "/etc/kube-flannel/net-conf.json"
+)
+
+type kubeSubnetManager struct {
+	client         clientset.Interface
+	nodeName       string
+	nodeStore      cache.StoreToNodeLister
+	nodeController *framework.Controller
+	subnetConf     *subnet.Config
+}
+
+func NewSubnetManager() (subnet.Manager, error) {
+	cfg, err := restclient.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize inclusterconfig: %v", err)
+	}
+	c, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize client: %v", err)
+	}
+
+	if os.Getenv("NODE_NAME") == "" {
+		return nil, fmt.Errorf("NODE_NAME not set")
+	}
+
+	var cniConf, netConf []byte
+	cniConf, err = ioutil.ReadFile(cniConfPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read cni conf: %v", err)
+	}
+	netConf, err = ioutil.ReadFile(netConfPath)
+	if err != nil {
```

Comment:
The even better solution would be that after flannel knows in which pod it runs, it could just fetch the values of it's annotations (with default values) and create a NetConf from those values.